add support for organizing imports at the bottom of file

tristanpemble opened the pull request:
this adds an `import_organization` config that allows `auto`, `top`, and `bottom`. it defaults to `auto`.

in auto mode, if the last decl of the file is an import, it uses `bottom`, otherwise it uses `top`.

I struggled with whitespace/newlines, as you can see in the test cases. I'm putting the PR up in the hope that either this is acceptable, since `zig fmt` will remove the duplicate newlines, or that if it is not acceptable, someone can assist me in fixing it.

tristanpemble commented:
this is now ready for review.

Techatrix commented on src/features/code_actions.zig:
```diff
@@ -587,6 +589,52 @@ fn handleVariableNeverMutated(builder: *Builder, loc: offsets.Loc) !void {
+fn analyzeImportPlacement(tree: Ast, imports: []const ImportDecl) ImportPlacement {
+    const root_decls = tree.rootDecls();
+
+    // If there are no declarations or imports, just return a default value
+    if (root_decls.len == 0 or imports.len == 0) return .top;
+
+    // Imports are not in source order, so we need to find the first and last import indices
+    var first_import: u32 = std.math.maxInt(u32);
+    var last_import: u32 = std.math.minInt(u32);
+    for (imports) |import| {
+        first_import = @min(first_import, @intFromEnum(import.var_decl));
+        last_import = @max(last_import, @intFromEnum(import.var_decl));
+    }
+    const first_decl: u32 = @intFromEnum(root_decls[0]);
+    const last_decl: u32 = @intFromEnum(root_decls[root_decls.len - 1]);
+
+    const starts_with_import = first_decl == first_import;
+    const has_gaps = root_decls.len != imports.len;
+    const ends_with_import = last_decl == last_import;
+
+    // These are the clear-cut cases.
+    if (starts_with_import and !ends_with_import) {
+        return .top;
+    } else if (ends_with_import and !starts_with_import) {
+        return .bottom;
+    } else if (!starts_with_import and !ends_with_import) {
+        return .top;
+    }
+
+    // In the ambiguous case, there is an import at both the top and bottom. The deciding factor is if there are other
+    // non-import declarations between them.
+    //
+    // If there aren't, the file is only imports, so we can choose "top" to avoid unnecessary newlines.
+    // If there are; there being an import at the bottom is a strong signal that that is what they want.
+    if (!has_gaps) {
+        return .top;
+    } else {
+        return .bottom;
+    }
+}
```
This function is more complicated than it needs to be:

- `imports` are already sorted.
- The usage `@intFromEnum` is unnecessary because you are comparing `Ast.Node.Index` which already works with `==` and `!=`.

```suggestion
fn analyzeImportPlacement(tree: Ast, imports: []const ImportDecl) ImportPlacement {
    const root_decls = tree.rootDecls();

    if (root_decls.len == 0 or imports.len == 0) return .top;

    const first_import = imports[0];
    const last_import = imports[imports.len - 1];

    const first_decl = root_decls[0];
    const last_decl = root_decls[root_decls.len - 1];

    const starts_with_import = first_decl == first_import.var_decl;
    const ends_with_import = last_decl == last_import.var_decl;

    if (starts_with_import and ends_with_import) {
        // In the ambiguous case, there is an import at both the top and bottom. The deciding factor is 
        // if there are other non-import declarations between them.
        //
        // If there aren't, the file is only imports, so we can choose "top" to avoid unnecessary newlines.
        // If there are; there being an import at the bottom is a strong signal that that is what they want.
        const has_gaps = root_decls.len != imports.len;
        return if (has_gaps) .bottom else .top;
    }
    return if (!starts_with_import and ends_with_import) .bottom else .top;
}
```

Techatrix reviewed (changes requested):
As mentioned in #2100, this feature should not have a config option. Remove it.

> I struggled with whitespace/newlines, as you can see in the test cases. I'm putting the PR up in the hope that either this is acceptable, since `zig fmt` will remove the duplicate newlines, or that if it is not acceptable, someone can assist me in fixing it.

Relying on `zig fmt` is acceptable. This could always be improved in the future.

tristanpemble commented:
Thank you for the feedback. I've removed the config option, and simplified `analyzeImportPlacement` as you've suggested.

The key was that I somehow completely missed the [sort call directly above](https://github.com/zigtools/zls/blob/af2acc31059f0ba2b1c81fc71a7e5569ef5500dd/src/features/code_actions.zig#L655) 🤦 . These checks needed to on the original index, not post-sort.

Anyways, thanks to your input, it's much cleaner now.
